### DISCUSSION TITLE
Make threads inherit allocator

### DISCUSSION THREAD

**Discussion by xzores** [odin-lang/Odin#2738] [Ideas/Requests]
I am currently using the tracking allocator, but it is actually almost useless to me as I use many threads, and getting it to work with many thread has been almost impossible for me.
It could help me to make threads inherit context.allocator from the thread creating it. I find this useful as it would then be possible to "contain" all the memory from a subsection of some code in the same allocator. This is not currently possible as if the sub-code uses thread, it would be required by the sub-code to handle setting an allocator if needed.

I know that making threads inherit allocators can be problematic because it then requires all the threads to share resources, and so I would like to suggest that an mem.Allocator also includes a copy_proc (or copy_to_thread_poc).
The copy_proc would  then be called when a new thread is started, the allocator can then decide to either make return itself or create a copy / new version of it self and pass to that thread.

Also I find it weird that you can set an allocator when using thread_pool, but when using thread_pool you cannot set a priority.
And when using a single thread you can set a priority, but not an allocator.

**Comment by xzores** [comment]
I can make a pull request, if this idea would be accepted-

**Comment by flysand7** [comment]
Looks like this is already working in Odin.

**Comment by xzores** [comment]
Not for threads_pools and temp allocators I belive (unless it changed since I tried)

**Comment by Tetralux** [comment]
You can do this via `t.init_context` where `t` is a thread returned from `thread.create`.
You can create a thread, then set that field, then call `thread.start`.
But do note that if you set that, then you must destroy the thread's temp allocator yourself, which is otherwise handled by `core:thread`.
(See [`core/thread/thread.odin#_maybe_destroy_default_temp_allocator`](https://github.com/odin-lang/Odin/blob/a06bde729bc92fe7f39ff79de1b9319bbf2c2bf6/core/thread/thread.odin#L315-L325).)
